[Kolors] Add PAG

asomoza opened the pull request:
# What does this PR do?

Adds PAG to kolors

Note: Kolors with PAG behaves different than with SDXL, I had to use a lower scale and different layers to get good results.

## How to test

### Txt2Img
```python
import torch

from diffusers import AutoPipelineForText2Image, DPMSolverMultistepScheduler


pipe = AutoPipelineForText2Image.from_pretrained(
    "Kwai-Kolors/Kolors-diffusers",
    torch_dtype=torch.float16,
    variant="fp16",
    enable_pag=True,
    pag_applied_layers=["down.blocks.2.attentions.1", "up.blocks.0.attentions.1"],
).to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config, use_karras_sigmas=True)

prompt = "cinematic film still of a pirate ship is navigating the milk inside of a steaming cup of coffee. The ship’s wooden hull glistens with droplets of coffee, and tiny pirates can be seen on deck, adjusting sails and peering through miniature telescopes. The tilt-shift effect adds a magical touch, high budget hollywood movie, cinemascope, epic, film grain"

image = pipe(
    prompt=prompt,
    negative_prompt="",
    guidance_scale=5.5,
    num_inference_steps=25,
    pag_scale=1.5,
).images[0]
 ```

|W/O|PAG|
|---|---|
|![20240721003413_227659598](https://github.com/user-attachments/assets/049c9737-7fe1-4373-acb6-1b8c6bb39380)|![20240722134235_227659598](https://github.com/user-attachments/assets/7baeca6e-b0c6-47ac-8ca8-37f2f6db42ad)|

After the PAG refactor, the image changed a bit, not sure why:

|before|after|
|---|---|
|![20240722134235_227659598](https://github.com/user-attachments/assets/3e0b3676-2bae-48c5-96d3-9143c404781f)|![20240806211616_227659598](https://github.com/user-attachments/assets/b9b19df6-bca9-4d49-9165-0cdd349b5f00)|

## Who can review?

@yiyixuxu @a-r-r-o-w @sunovivid @sayakpaul

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8934). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

yiyixuxu commented on src/diffusers/pipelines/pag/pipeline_pag_kolors.py:
```diff
@@ -0,0 +1,1130 @@
+        >>> import torch
+        >>> from diffusers import KolorsPipeline
+
+        >>> pipe = KolorsPipeline.from_pretrained(
```
update this?

yiyixuxu commented on src/diffusers/pipelines/pag/pipeline_pag_kolors.py:
```diff
@@ -0,0 +1,1130 @@
+
+        self.set_pag_applied_layers(pag_applied_layers)
+
+    def encode_prompt(
```
copied from?

yiyixuxu commented on src/diffusers/pipelines/pag/pipeline_pag_kolors.py:
```diff
@@ -0,0 +1,1130 @@
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
copied from?

yiyixuxu reviewed (approved):
thanks!
add a test?

asomoza commented:
Added the fast tests and I fixed the old ones because I needed to get the original Kolors Pipeline to test the AutoPipeline. In short, I had to add all the missing tests I had as TODO for kolors.

@yiyixuxu @sayakpaul should I add the slow tests for PAG? this one has a really big text encoder which is not in transformers yet.

sayakpaul commented:
> @yiyixuxu @sayakpaul should I add the slow tests for PAG? this one has a really big text encoder which is not in transformers yet.

I would say, it'd be okay skipping the SLOW tests for now.

a-r-r-o-w commented:
We should try and merge this before next release, no?

sayakpaul commented on src/diffusers/pipelines/kolors/tokenizer.py:
```diff
@@ -143,10 +143,18 @@ def get_command(self, token):
     def unk_token(self) -> str:
         return "<unk>"
 
+    @unk_token.setter
+    def unk_token(self, value: str):
+        self._unk_token = value
```
Seems like these should have been done in a separate PR?

asomoza replied:
if I don't do this, the tests fails (throws AttributeError), I'm ok with doing this in a separate PR but then I'll have to skip the tests in this PR

sayakpaul replied:
Ah okay. Then no worries.

sayakpaul commented on tests/pipelines/kolors/test_kolors.py:
```diff
@@ -133,23 +133,11 @@ def test_inference(self):
-    # not sure if it is worth to fix it before integrating it to transformers
     def test_save_load_optional_components(self):
-        # TODO (Alvaro) need to fix later
-        pass
+        super().test_save_load_optional_components(expected_max_difference=2e-4)
```
Same. Probably should have been done in a separate PR.

sayakpaul reviewed (approved):
LGTM. Let's add the pipeline to the list of pipelines we support in the [docs](https://huggingface.co/docs/diffusers/main/en/api/pipelines/pag)?
